(26153) Version-based download/build of cvmfsexec distribution tar files

namrathaurs opened the pull request:
First version of the logic for version-based download/build of cvmfsexec distribution tar files. Testing of the logic in progress (therefore, draft PR).

namrathaurs committed:
for ticket 26153 - cvmfsexec versioning logic

- initial logic for checking cvmfsexec version to determine if the distribution tarballs need to be rebuilt during the factory reconfig and/or upgrade

namrathaurs committed:
versioning logic optimized after testing for various scenarios

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
-    mkdir -p $cvmfsexec_tarballs
-fi
+    # build the distributions for cvmfsexec based on the source, os and platform combination
+    cvmfsexec_distros=$cvmfsexec_temp/distros
+    if [[ ! -d $cvmfsexec_distros ]]; then
+        mkdir -p $cvmfsexec_distros
```
Quote paths

namrathaurs replied:
updated

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
+            echo -n "Making $cvmfs_src distribution for $mach_type machine..."
+            os=`echo $mach_type | awk -F'-' '{print $1}'`
+            arch=`echo $mach_type | awk -F'-' '{print $2}'`
+            $cvmfsexec_latest/makedist -m $mach_type $cvmfs_src &> /dev/null
+            if [[ $? -eq 0 ]]; then
```
Better in one line and remember to quote the paths: `if "$cvmfsexec_latest"/makedist -m $mach_type $cvmfs_src &> /dev/null ; then`

namrathaurs replied:
updated

mambelli reviewed (changes requested):
All paths that could have spaces should be quoted. I highlighted only some.

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
+fi
+
+# first, checking in the work-dir location for the current version of cvmfsexec
 work_dir=$(grep -m 1 "submit" /etc/gwms-factory/glideinWMS.xml | awk -F"\"" '{print $6}')
```
created a variable `factory_config_file` and initialized the file path to the factory config file for use in the `work_dir=...` line

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
+# first, checking in the work-dir location for the current version of cvmfsexec
 work_dir=$(grep -m 1 "submit" /etc/gwms-factory/glideinWMS.xml | awk -F"\"" '{print $6}')
 cvmfsexec_tarballs=$work_dir/cvmfsexec/tarballs
+if [[ -d $cvmfsexec_tarballs && -f "$cvmfsexec_tarballs/.cvmfsexec_version" ]]; then
```
updated

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
-    fi
+# otherwise, .cvmfsexec_version file does not exist from a previous upgrade or it's a first-time factory upgrade
+# check if the temp directory for cvmfsexec from previous run exists
+ls -d $sys_tmp/cvmfsexec.*/ > /dev/null 2>&1
```
updated

namrathaurs committed:
changes addressing PR review comments

- added a variable to store the location of factory configuration XML
- added protection aginst non-existence of cvmfsexec directory (fresh install of GWMS with first run of factory upgrade)
- quoted variables containing paths to include correct referencing of file paths with whitespaces, when applicable
- eliminated the use of mktemp/cvmfsexec.XXX directory name pattern since the cvmfsexec temp directory will be reused across multiple invocations
- optimized use of exit status of command in if conditions
- added logic for removal of distros/ and latest/ folder after the distributions are built

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,103 @@ CVMFS_SOURCES=osg:egi:default
+if [[ $? -eq 0 ]]; then
+    # cvmfsexec.XXX (temp) directory exists; reuse this
+    cvmfsexec_temp=$(ls -d $sys_tmp/cvmfsexec.*/)
+    cvmfsexec_temp=${cvmfsexec_temp%?}   # removes the trailing '/' to avoid incorrect paths later
```
the new commit makes use of static name `cvmfsexec.tmp` and is created under `work-dir` location as suggested.

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,113 @@ CVMFS_SOURCES=osg:egi:default
-machine_types=($(echo $SUPPORTED_TYPES | tr ":" "\n"))
+    # build the distributions for cvmfsexec based on the source, os and platform combination
+    cvmfsexec_distros="$cvmfsexec_temp"/distros
+    if [[ ! -d "$cvmfsexec_distros" ]]; then
+        mkdir -p "$cvmfsexec_distros"
+    fi
+
+    if [[ ! -d "$cvmfsexec_tarballs" ]]; then
+        mkdir -p "$cvmfsexec_tarballs"
+    fi
```
`mkdir -p` is OK if the directory is already there. Does nothing, no need to check.
I think you created already above the tarballs directory

namrathaurs replied:
Correct, the tarballs directory will be created at the beginning. Removed lines 86:88 as suggested.

namrathaurs committed:
removed the check for system temp directory

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,113 @@ CVMFS_SOURCES=osg:egi:default
-    echo "cvmfsexec version: `$cvmfsexec_base/cvmfsexec -v`"
+factory_config_file="/etc/gwms-factory/glideinWMS.xml"
 
+sys_tmp=$(echo $TMPDIR)
+if [[ -z "$sys_tmp" ]]; then
+    sys_tmp=/tmp
 fi
```
removed

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,113 @@ CVMFS_SOURCES=osg:egi:default
+    # if the cvmfsexec directory does not exist
+    mkdir -p $work_dir/cvmfsexec
+    # create a directory named tarballs under cvmfsexec directory
+    cvmfsexec_tarballs=$(mkdir -p $work_dir/cvmfsexec/tarballs -v | awk -F" " '{print $4}' | sed 's/^.//; s/.$//' )
```
Thank you for the notes on best practices.

> * be as specific as possible. If you have to remove a quote, remove a quote, not the first character whatever it is (`sed "s/^'//; s/'$//"` -may need to quote, not sure-), even better bash builtins avoid the invocation of a subshell:
> 
> ```
> cvmfsexec_tarballs="${cvmfsexec_tarballs#'}"
> cvmfsexec_tarballs="${cvmfsexec_tarballs%'}"
> ```
I first tried with the sed pattern that you suggested, which is to remove the quote character rather than the first or last character. However, the kind of quote character in the output of `mkdir -v` was not being matched with the pattern as it was a different style of quotes. This was the sole reason I resorted to removing the first and last character.


> Then I do not understand the reason of the pipe, it will return "$work_dir/cvmfsexec/tarballs" if all is OK

I used the pipe to be able to store the location of the directory that was newly created instead of initializing with `$work-dir/cvmfsexec/tarballs`.

I like the suggestion of including a check for failure of directory creation. I've incorporated the same in one of the new commits.

namrathaurs commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -17,86 +17,113 @@ CVMFS_SOURCES=osg:egi:default
-    mkdir -p $cvmfsexec_tarballs
+# otherwise, .cvmfsexec_version file does not exist from a previous upgrade or it's a first-time factory upgrade
+# check if the temp directory for cvmfsexec from previous run exists
+if ls -d $work_dir/cvmfsexec/cvmfsexec.tmp > /dev/null 2>&1 ; then
+    # cvmfsexec temp directory exists; reuse this
+    cvmfsexec_temp=$(ls -d $work_dir/cvmfsexec/cvmfsexec.tmp)
+else
+    # cvmfsexec temp directory does not exist; so create one
+    cvmfsexec_temp=$(mkdir -p $work_dir/cvmfsexec/cvmfsexec.tmp -v | awk -F" " '{print $4}' | sed 's/^.//; s/.$//' )  # remove the single quotes at the start and end of the string
+    chmod 755 "$cvmfsexec_temp"
 fi
```
updated

namrathaurs committed:
suggested changes as per second round of PR review
